src/planner/binder/tableref/bind_basetableref.cpp

```diff
@@ -45,9 +45,11 @@ unique_ptr<BoundTableRef> Binder::Bind(BaseTableRef &ref) {
 			return Bind(subquery, found_cte);
 		} else {
 			// There is a CTE binding in the BindContext.
-			// This can only be the case if there is a recursive CTE present.
+			// This can only be the case if there is a recursive CTE,
+			// or a materialized CTE present.
 			auto index = GenerateTableIndex();
-			auto result = make_uniq<BoundCTERef>(index, ctebinding->index);
+			auto result = make_uniq<BoundCTERef>(index, ctebinding->index,
+			                                     cte.materialized == CTEMaterialize::CTE_MATERIALIZE_ALWAYS);
```

Comment:
As materialized and non-materialized CTEs should always have the same result, perhaps it makes sense to switch over the default when alternative verification is enabled. For example:

```cpp
auto materialized = cte.materialized;
if (materialized == CTEMaterialize::CTE_MATERIALIZE_DEFAULT) {
#ifdef DUCKDB_ALTERNATIVE_VERIFY
   materialized = CTEMaterialize::CTE_MATERIALIZE_ALWAYS;
#else
   materialized = CTEMaterialize::CTE_MATERIALIZE_NEVER;
#endif
}
...
```

That way when compiling with `ALTERNATIVE_VERIFY=1 make` all CTEs will be materialized by default. This will enable some free extra testing.

Reply:
In principle, you are right—and I like the idea. However, there is a corner case, where `MATERIALIZED` CTEs behave differently:

```
WITH RECURSIVE t(x) AS (
  SELECT 1
    UNION ALL
  SELECT x+1
  FROM   t
)
SELECT * FROM t LIMIT 4;
```

This query expresses an infinite loop. It only stops, because of the `LIMIT` clause.
If you materialize this CTE, it does *not* terminate anymore.

Reply:
That makes sense. I imagine there are only a few tests where those types of CTEs are run, though. Perhaps we could add a new option (e.g. `noalternateverify`) to those tests - and add `require noalternateverify` to them so the materialized CTE test is skipped?

Reply:
Yes, there are a few tests, but not many. Adding such an option sounds good to me. I will look into it.

Reply:
I've added the new option `noalternativeverify` to the unittest and annotated the relevant tests with it.